turbopack/crates/turbo-tasks-fs/src/watcher.rs

```diff
@@ -131,8 +136,28 @@ impl NonRecursiveDiskWatcherState {
     ) -> Result<()> {
         if self.watching.contains(dir_path) {
             let mut internal = watcher.internal.lock().unwrap();
-            // TODO: Also restore any watchers for children of this directory
-            self.start_watching_dir(&mut internal, dir_path, root_path)?;
+
+            // watch the new directory and any of it's parents
+            self.start_watching_dir(
+                &mut internal,
+                dir_path,
+                root_path,
+                /* watch_parents */ true,
+            )?;
+
+            // Also try to restore any watchers for children of this directory
+            for child_path in self
+                .watching
+                .iter()
+                .filter(|p| p.key().starts_with(dir_path))
+            {
+                self.start_watching_dir(
+                    &mut internal,
+                    child_path.key(),
+                    root_path,
+                    /* watch_parents */ false,
```

Comment:
Maybe leave a note here that this flag is important to avoid the deadlock on `watching` as we can't modify the `watching` set while iterating it.